compiler/rustc_passes/src/check_attr.rs

```diff
@@ -1131,6 +1131,39 @@ impl CheckAttrVisitor<'_> {
             }
         }
 
+        if !matches!(
+            target,
+            Target::Fn
+                | Target::Enum
+                | Target::Struct
+                | Target::Union
+                | Target::Method(_)
+                | Target::ForeignFn
+                // FIXME(patchmixolydic): this doesn't seem to work on user code,
+                // but seems to work for certain traits in core (eg. `Future`).
+                // maybe it's a lang item thing?
+                | Target::Trait
```

Comment:
Example ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=870d5c94b6ca6dd06a76083770be59f7)):
```rust
#[must_use = "lol"]
trait Foo {
    fn bar(&self) -> usize;
}

impl Foo for Vec<usize> {
    fn bar(&self) -> usize {
        self.len()
    }
}

fn main() {
    vec![1usize];
    vec![1usize].bar();
    async {};
}
```
```
warning: unused implementer of `Future` that must be used
  [--> src/main.rs:15:5
](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021#)   |
15 |     async {};
   |     ^^^^^^^^^
   |
   = note: `#[warn(unused_must_use)]` on by default
   = note: futures do nothing unless you `.await` or poll them
```

Reply:
must_use on the trait only warns for opaque types like `impl Trait`, not for concrete types: [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=5d7e9b573954d7f04029e14ce9a03444)

Reply:
Thanks! I probably should've checked [the Reference](https://doc.rust-lang.org/reference/attributes/diagnostics.html#the-must_use-attribute) :sweat_smile: